### ISSUE TITLE
`jsx-handler-names` not working 🤔 

### ISSUE THREAD

**Issue by mkosir** [jsx-eslint/eslint-plugin-react#3348]
Not sure what I'm doing wrong but no matter how I set/enable `jsx-handler-names` rule in my projects (Nextjs) it doesn't seem to be working at all 🤔 Any idea what to check?
```js
"extends": [
    "plugin:react/recommended",
...
"rules": {
    "react/jsx-handler-names": "error",
...
```

**Comment by ljharb** [comment]
Can you provide some example code that you expect to be warned, but isn't?

**Comment by mkosir** [comment]
Hey @ljharb,
with `create-next-app` I initialized an [empty project](https://github.com/mkosir/example-lint) and added [`jsx-handler-names` lint rule](https://github.com/mkosir/example-lint/blob/main/.eslintrc#L15) to it.
If you run `npm run lint` linting will succeed without errors, with [example component](https://github.com/mkosir/example-lint/blob/main/components/TestComponent.tsx) where linting should fail.
Could it be that this rule only works with (old) class based React components, but don't see it being mentioned in the docs 🤔

**Comment by ljharb** [comment]
@mkosir instead of just `"react/jsx-handler-names": "error"`, try `"react/jsx-handler-names": ["error", { checkLocalVariables: true }]` - with that setting, i'm able to see the expected errors.

**Comment by ljharb** [comment]
Props should always be destructured; it’s just that this rule’s defaults are waiting for a semver-major release to change.

**Comment by mkosir** [comment]
> Props should always be destructured

Hm not following completely 🤔 

I updated the [test repo](https://github.com/mkosir/example-lint) mentioned above, if you try to run `npm run lint` linting fails on [this line](https://github.com/mkosir/example-lint/blob/main/components/TestComponent.tsx#L20) (destructed callback prop `onClick={onShouldLint}`), but it shouldn't.
Where if we [don't destruct props](https://github.com/mkosir/example-lint/blob/main/components/TestComponent.tsx#L36), lint rule works as expected.

**Comment by ljharb** [comment]
Yes, I understand. The takeaway is that the `jsx-handler-names` rule should always be used with the `checkLocalVariables` option, it's just that this plugin can't change the default without it being a breaking change.

**Comment by mkosir** [comment]
Yep it make sense to change rule defaults 👍 
Thx for clarification on this one, unfortunately I wont be enabling this (otherwise awesome) rule, since props destructing is must have.

**Comment by ljharb** [comment]
Again, you can enable it *with the checkLocalNames option*, and it will work just fine with destructuring.

**Comment by mkosir** [comment]
> you can enable it with the checkLocalNames option

You probably ment `checkLocalVariables` 🤔 ?
If thats the case I have already [enabled it](https://github.com/mkosir/example-lint/blob/main/.eslintrc#L15) and [lint fails](https://github.com/mkosir/example-lint/blob/main/components/TestComponent.tsx#L20) as mentioned (but it shouldn't)

**Comment by ljharb** [comment]
Yes, it should; handlers must start with “handle” - starting with “on” is for prop names.

**Comment by mkosir** [comment]
Hm but [onShouldLint](https://github.com/mkosir/example-lint/blob/main/components/TestComponent.tsx#L20) is a prop name just destructed.
Anyhow I just feel I own you a 🍺 IRL for the discussion I'm putting you through 😄

**Comment by ljharb** [comment]
ah, i see what you mean. You’d want to rename the prop while destructuring it.
